### ISSUE TITLE
`insertOrUpdate` for `PostgresProfile` generates syntactically wrong  SQL

### ISSUE THREAD

**Issue by jkugiya** [slick/slick#2207]
**DESCRIPTION:**
`insertOrUpdate` for `PostgresProfile` and `DerbyProfile` generates syntactically wrong  SQL.

**STEPS TO REPRODUCE:**

Given table like this:
```scala
    case class V(id: Int, value: String)
    // all columns are primary key
    class T2(tag: Tag) extends Table[V](tag, "mytable2") {
      def id = column[Int]("id")
      def value = column[String]("value", O.SqlType("varchar(20)"))
      def * = (id, value) <> ((V.apply _).tupled, V.unapply)
      def pk = primaryKey("pk", (id, value)) // multiple primary key
    }
```

```scala
val ts2 = TableQuery[T2]

println(ts2.insertOrUpdate(V(0, "c")).statements.mkString("\n"))

```
**EXPECTED RESULT:**
Generates right set clause for UPDATE statement or support native upsert statement.

**ACTUAL RESULT:**

```sql
-- Set clause in the UPDATE statement is wrong
update "mytable2" set  where "id"=? and "value"=?; insert into "mytable2" ("id","value") select ?,? where not exists (select 1 from "mytable2" where "id"=? and "value"=?)
```

**RELATED TO**
- https://github.com/slick/slick/issues/966#issuecomment-379232820
  - Similar, but it seems to be different.
- #2206
  - `DerbyProfile` Issue that occurs under the same conditions

**Comment by nafg** [comment]
I'm confused what is the use case, this doesn't sound like "upsert" but "insert if not exists" since there are no columns to _update_

**Comment by jkugiya** [comment]
At the moment, I have not dug deep into why it generates such SQL, but some SQL profiles that do not directly support upsert in the database do separate insert and update statements depending on the result of the select statement. Perhaps in this case, that feature is not working well.

https://github.com/slick/slick/blob/06ccee3cdc0722adeb8bb0658afb4a0d3524b119/slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala#L588-L607

**Comment by nafg** [comment]
Exactly. If you Ctrl-click through that enough times (from `updateInsert`) you get to this code (which is duplicated in PostgresProfile separately):

```scala
"update " + tableName + " set " + softNames.map(n => s"$n=?").mkString(",") + " where " + pkNames.map(n => s"$n=?").mkString(" and ")
```

If `softNames` is empty then you get an empty `SET` clause.

The problem is (AFAICT) as things are currently designed there's no obvious way for the code that deep to signal to `emulate` _not to run **any** statement_.

That said I'm not fully convinced this is expected work. In #2503 I added a `Capability` to tell whether "upsert as 'insert-if-not-exists'" is supported.

My motivation is that otherwise `slick-testkit-example` (and possibly other custom profiles in the wild) fails because it doesn't match `case _: PostgresProfile | _: DerbyProfile =>` in `InsertTests#testInsertOrUpdateWithPrimaryKeyOnly`. As I said using `confName` is surely not robust. Keep in mind that TestKit is not just a fancy test suite for Slick. It's meant to enable others to write custom profiles that work as Slick expects them to.

**Comment by nafg** [comment]
Do you see what I mean about this use case really not being "insertOrUpdate" but rather "insert if not exists"? Do you agree that this is not a bug report, but a nice feature request?

Although at least it should report the error better, I suppose.

**Comment by jkugiya** [comment]
It is certainly not clear which rows should be updated with the update or upsert statement when the column only has a compound primary key. If anything, the problem may be that the behavior in that case is not constant.

> Do you see what I mean about this use case really not being "insertOrUpdate" but rather "insert if not exists"? > Do you agree that this is not a bug report, but a nice feature request?

I see for now. I understand that an error in that test is not a bug. However, allowing for different post-conditions for different database products may cause confusion.

**Comment by jkugiya** [comment]
> Do you see what I mean about this use case really not being "insertOrUpdate" but rather "insert if not exists"?

Isn't "insert if not exists" part of the general behavior of `insertOrUpdate`? I believe the error is not caused by the inability to identify the row to be updated, but by the generation of an invalid `set` clause. I think this means that the particular approach of generating an `update` statement at the beginning is not very reasonable.

**Comment by jkugiya** [comment]
> generating an update statement at the beginning is not very reasonable.

Why this approach is being taken is something that needs to be investigated.

**Comment by nafg** [comment]
> > Do you see what I mean about this use case really not being "insertOrUpdate" but rather "insert if not exists"?
> 
> Isn't "insert if not exists" part of the general behavior of `insertOrUpdate`? I believe the error is not caused by the inability to identify the row to be updated, but by the generation of an invalid `set` clause. I think this means that the particular approach of generating an `update` statement at the beginning is not very reasonable.

Yes it's  _part_ of it. When you say insertOrUpdate one could argue that's explicitly saying "but if it does exist then do an update."

Separate question is why postgres is using emulated...

As for the internals I think it's just an artifact of Slick's general overengineering.

**Comment by jkugiya** [comment]
I don't have time to investigate the commit logs on weekdays, so this is just my opinion at the moment, but perhaps it would be better to branch the DML to be issued according to the result of the select statement, as is the case with emulation in other databases.

As for postgres, the new version(15) seems to support merge insert, but it should be optional.
